# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/platforms/apple/AppleVTDecoder.cpp
# Commit: eb882d1719d2
# Full Hash: eb882d1719d2db8d44a48599596a5aedbe899c69
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2020-05-15 21:52:36
# Description:
#   Bug 1636615: Work around Apple VT decoder saying it dropped a frame, and then decoding that frame anyway. r=jya
#   
#   We do this by keeping the unexpected frame in mReorderQueue, and outputting it later when the caller is expecting frames.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D75116
# ==============================================================================

diff -r 1b2022e31cd9 -r eb882d1719d2 dom/media/platforms/apple/AppleVTDecoder.cpp
--- a/dom/media/platforms/apple/AppleVTDecoder.cpp	Fri May 15 13:56:47 2020 +0000
+++ b/dom/media/platforms/apple/AppleVTDecoder.cpp	Thu May 14 23:45:26 2020 +0000
@@ -188,7 +188,7 @@
     MonitorAutoLock mon(mMonitor);
     // Smile and nod
     NS_WARNING("Decoder synchronously dropped frame");
-    mPromise.ResolveIfExists(DecodedData(), __func__);
+    MaybeResolveBufferedFrames();
     return;
   }
 
@@ -303,6 +303,20 @@
   decoder->OutputFrame(image, *frameRef);
 }
 
+void AppleVTDecoder::MaybeResolveBufferedFrames() {
+  mMonitor.AssertCurrentThreadOwns();
+
+  if (mPromise.IsEmpty()) {
+    return;
+  }
+
+  DecodedData results;
+  while (mReorderQueue.Length() > mMaxRefFrames) {
+    results.AppendElement(mReorderQueue.Pop());
+  }
+  mPromise.Resolve(std::move(results), __func__);
+}
+
 // Copy and return a decoded frame.
 void AppleVTDecoder::OutputFrame(CVPixelBufferRef aImage,
                                  AppleVTDecoder::AppleFrameRef aFrameRef) {
@@ -321,7 +335,7 @@
     // Image was dropped by decoder or none return yet.
     // We need more input to continue.
     MonitorAutoLock mon(mMonitor);
-    mPromise.Resolve(DecodedData(), __func__);
+    MaybeResolveBufferedFrames();
     return;
   }
 
@@ -433,11 +447,7 @@
   // in composition order.
   MonitorAutoLock mon(mMonitor);
   mReorderQueue.Push(data);
-  DecodedData results;
-  while (mReorderQueue.Length() > mMaxRefFrames) {
-    results.AppendElement(mReorderQueue.Pop());
-  }
-  mPromise.Resolve(std::move(results), __func__);
+  MaybeResolveBufferedFrames();
 
   LOG("%llu decoded frames queued",
       static_cast<unsigned long long>(mReorderQueue.Length()));